xdpderp/xdp.c

```diff
@@ -0,0 +1,513 @@
+		return end(ctx, XDP_PASS, is_ipv6);
+	}
+	if (attr_data + 8 > data_end) {
+		return end(ctx, XDP_PASS, is_ipv6);
+	}
+	char want_sw[] = {0x74, 0x61, 0x69, 0x6c, 0x6e, 0x6f, 0x64, 0x65}; // tailnode
+	char *got_sw = attr_data;
+	for (int j = 0; j < 8; j++) {
+		if (got_sw[j] != want_sw[j]) {
+			return end(ctx, XDP_PASS, is_ipv6);
+		}
+	}
+
+	// Begin transforming packet into a STUN_BINDING_RESPONSE. From here
+	// onwards we return XDP_ABORTED instead of XDP_PASS when transformations or
+	// bounds checks fail as it would be nonsensical to pass a mangled packet
+	// through to the kernel, and we may be interested in debugging via
+	// tracepoint.
+
+	// Set success response and new length. Magic cookie and txid remain the
+	// same.
+	req->type = bpf_htons(STUN_BINDING_RESPONSE);
+	if (is_ipv6) {
+		req->length = bpf_htons(sizeof(struct stunattr) + sizeof(struct stunxor6));
+	} else {
+		req->length = bpf_htons(sizeof(struct stunattr) + sizeof(struct stunxor));
+	}
+
+	// Set attr type. Length remains unchanged, but set it again for future
+	// safety reasons.
+	sa->num = bpf_htons(STUN_ATTR_XOR_MAPPED_ADDR);
+	if (is_ipv6) {
+		sa->length = bpf_htons(sizeof(struct stunxor6));
+	} else {
+		sa->length = bpf_htons(sizeof(struct stunxor));
+	}
+
+	// Set attr data.
+	struct stunxor *xor;
+	struct stunxor6 *xor6;
+	if (is_ipv6) {
+		xor6 = attr_data;
+		if ((void *)(xor6 + 1) > data_end) {
+			int expand = (void *)(xor6 + 1) - data_end;
+			if (bpf_xdp_adjust_tail(ctx, expand)) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			data_end = (void *)(long)ctx->data_end;
+			data = (void *)(long)ctx->data;
+			eth = data;
+			if ((void *)(eth + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			ip6 = (void *)(eth + 1);
+			if ((void *)(ip6 + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			udp = (void *)(ip6 + 1);
+			if ((void *)(udp + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			req = (void *)(udp + 1);
+			if ((void *)(req + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			sa = (void *)(req + 1);
+			if ((void *)(sa + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+			xor6 = (void *)(sa + 1);
+			if ((void *)(xor6 + 1) > data_end) {
+				return end(ctx, XDP_ABORTED, is_ipv6);
+			}
+		}
+		xor6->unused = 0x00; // unused byte
+		xor6->family = 0x02;
+		xor6->port = bpf_htons(bpf_ntohs(udp->source) ^ STUN_MAGIC_FOR_PORT_XOR);
+		xor6->addr[0] = bpf_htonl(bpf_ntohl(ip6->saddr.in6_u.u6_addr32[0]) ^ STUN_MAGIC);
+		for (int i = 1; i < 4; i++) {
+			// All three are __be32, no endianness flips.
+			xor6->addr[i] = ip6->saddr.in6_u.u6_addr32[i] ^ req->txid[i-1];
+		}
+	} else {
+		xor = attr_data;
+		if ((void *)(xor + 1) > data_end) {
+			return end(ctx, XDP_ABORTED, is_ipv6);
+		}
+		xor->unused = 0x00; // unused byte
+		xor->family = 0x01;
+		xor->port = bpf_htons(bpf_ntohs(udp->source) ^ STUN_MAGIC_FOR_PORT_XOR);
```

Comment:
I think `udp->source ^ bpf_htons(STUN_MAGIC >> 16)` is equivalent

Reply:
updated in https://github.com/tailscale/tailscale/pull/12390/commits/8383d4a490154bec46a6eb026a9ac0a88f0fd14e